src/providers/hana/qgshanaconnection.cpp

```diff
@@ -0,0 +1,621 @@
+{
+  if ( layerProperty.geometryColName.isEmpty() )
+    return -1;
+
+  int ret = -1;
+
+  try
+  {
+    StatementRef stmt = mConnection->createStatement();
+
+    if ( !layerProperty.isView )
+    {
+      QString sql = QStringLiteral( "SELECT SRS_ID FROM SYS.ST_GEOMETRY_COLUMNS "
+                                    "WHERE SCHEMA_NAME = '%1' AND TABLE_NAME = '%2' AND COLUMN_NAME = '%3'" )
+                    .arg( layerProperty.schemaName, layerProperty.tableName, layerProperty.geometryColName );
+      ResultSetRef rsSrid = stmt->executeQuery( reinterpret_cast<const char16_t *>( sql.unicode() ) );
+      ret = rsSrid->next() ? *rsSrid->getInt( 1 ) : -1;
+      rsSrid->close();
+    }
+    else
+    {
+      QString sql = QStringLiteral( "SELECT %1.ST_SRID() FROM %2.%3 WHERE %1 IS NOT NULL LIMIT %4" )
+                    .arg( QgsHanaUtils::quotedIdentifier( layerProperty.geometryColName ),
+                          QgsHanaUtils::quotedIdentifier( layerProperty.schemaName ),
+                          QgsHanaUtils::quotedIdentifier( layerProperty.tableName ),
+                          QString::number( GEOM_TYPE_SELECT_LIMIT ) );
+      ResultSetRef rsSrid = stmt->executeQuery( reinterpret_cast<const char16_t *>( sql.unicode() ) );
+      int srid = -1, prevSrid = -1;
+      while ( rsSrid->next() )
+      {
+        srid = *rsSrid->getInt( 1 );
+        if ( prevSrid != -1 && srid != prevSrid )
+        {
+          srid = -1;
+          break;
+        }
+        prevSrid = srid;
+      }
+
+      ret = srid;
+      rsSrid->close();
+    }
+  }
+  catch ( const Exception &ex )
+  {
+    throw QgsHanaException( ex.what() );
+  }
+
+  return ret;
+}
+
+QStringList QgsHanaConnection::getLayerPrimaryeKeys( const QgsHanaLayerProperty &layerProperty )
+{
+  QStringList ret;
+
+  try
+  {
+    DatabaseMetaDataRef dbmd = mConnection->getDatabaseMetaData();
+    ResultSetRef rsPrimaryKeys = dbmd->getPrimaryKeys( nullptr,
+                                 layerProperty.schemaName.toStdString().c_str(),
+                                 layerProperty.tableName.toStdString().c_str() );
+
+    while ( rsPrimaryKeys->next() )
+    {
+      auto keyName = rsPrimaryKeys->getNString( 4 );
+      if ( !keyName.isNull() )
+      {
+        QString clmName = QgsHanaUtils::toQString( keyName );
+        ResultSetRef rsColumns = dbmd->getColumns( nullptr,
+                                 layerProperty.schemaName.toStdString().c_str(),
+                                 layerProperty.tableName.toStdString().c_str(),
+                                 clmName.toStdString().c_str() );
+
+        if ( rsColumns->next() )
+        {
+          Short dataType = rsColumns->getShort( 5 );
+          if ( !dataType.isNull() )
+          {
+            short dt = *dataType;
+            if ( dt == SQLDataTypes::TinyInt || dt == SQLDataTypes::SmallInt ||
+                 dt == SQLDataTypes::Integer || dt == SQLDataTypes::BigInt )
+              ret << clmName;
```

Comment:
What about non-integer primary keys or non-integer primary key parts?

Reply:
As far as I know, non-integer primary keys are not supported by QGIS.